daemon/k8s_watcher.go

```diff
@@ -850,6 +853,102 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumEndpoint, metricCreate, valid, equal) }()
+						if ciliumEndpoint, ok := obj.(*types.CiliumEndpoint); ok {
+							valid = true
+							endpoint := ciliumEndpoint.DeepCopy()
+							serCiliumEndpoints.Enqueue(func() error {
+								endpointUpdated(endpoint)
+								d.K8sEventProcessed(metricCiliumEndpoint, metricCreate, true)
+								return nil
+							}, serializer.NoRetry)
+						}
+					},
+					UpdateFunc: func(oldObj, newObj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumEndpoint, metricUpdate, valid, equal) }()
+						if ciliumEndpoint, ok := newObj.(*types.CiliumEndpoint); ok {
+							valid = true
+							endpoint := ciliumEndpoint.DeepCopy()
+							serCiliumEndpoints.Enqueue(func() error {
+								endpointUpdated(endpoint)
+								d.K8sEventProcessed(metricCiliumEndpoint, metricUpdate, true)
+								return nil
+							}, serializer.NoRetry)
+						}
+					},
+					DeleteFunc: func(obj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumEndpoint, metricDelete, valid, equal) }()
+						deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
+						if ok {
+							// Delete was not observed by the
+							// watcher but is removed from
+							// kube-apiserver. This is the last
+							// known state and the object no longer
+							// exists.
+							if ciliumEndpoint, ok := deletedObj.Obj.(*types.CiliumEndpoint); ok {
+								endpoint := ciliumEndpoint.DeepCopy()
+								serCiliumEndpoints.Enqueue(func() error {
+									endpointDeleted(endpoint)
+									return nil
+								}, serializer.NoRetry)
+								return
+							}
+						} else if ciliumEndpoint, ok := obj.(*types.CiliumEndpoint); ok {
+							endpoint := ciliumEndpoint.DeepCopy()
+							serCiliumEndpoints.Enqueue(func() error {
+								endpointDeleted(endpoint)
+								return nil
+							}, serializer.NoRetry)
+							return
+						}
+					},
+				},
+				k8s.ConvertToCiliumEndpoint,
+			)
+			isConnected := make(chan struct{})
+			// once isConnected is closed, it will stop waiting on caches to be
+			// synchronized.
+			d.blockWaitGroupToSyncResources(isConnected, ciliumEndpointInformer, k8sAPIGroupCiliumEndpointV2)
+
+			once.Do(func() {
+				// Signalize that we have put node controller in the wait group
+				// to sync resources.
+				asyncControllers.Done()
+			})
+			d.k8sAPIGroups.addAPI(k8sAPIGroupCiliumEndpointV2)
+			go ciliumEndpointInformer.Run(isConnected)
+
+			<-kvstore.Client().Connected()
+			close(isConnected)
+
+			log.Info("Connected to key-value store, stopping CiliumEndpoint watcher")
+
+			d.k8sAPIGroups.removeAPI(k8sAPIGroupCiliumEndpointV2)
+			// Create a new node controller when we are disconnected with the
+			// kvstore
+			<-kvstore.Client().Disconnected()
+
+			log.Info("Disconnected from key-value store, restarting CiliumEndpoint watcher")
```

Comment:
Should this code be protected by the `disable-endpoint-crd` option, to ensure that we don't cause trouble in a high-scale cluster whenever kvstore connectivity hits a blip?

Reply:
Do you mean to not even attempt to install the watcher in the first place if `disable-endpoint-crd` is set?

Reply:
I see that we are already automatically enabling the endpoint CRD when the kvstore is disabled so we should be all good?

Reply:
Right, I was suggesting to avoid this watcher codepath entirely if users are running with endpoint CRDs disabled. If that's already the case then feel free to ignore the suggestion.